src/unix/async.c

```diff
@@ -62,33 +69,30 @@ int uv_async_send(uv_async_t* handle) {
     return 0;
 
   /* Tell the other thread we're busy with the handle. */
+  if (cmpxchgi(&handle->u.fd, 0, 1) != 0)
+    return 0;
+
+  /* Set async pending  */
   if (cmpxchgi(&handle->pending, 0, 1) != 0)
     return 0;
 
   /* Wake up the other thread's event loop. */
   uv__async_send(handle->loop);
 
   /* Tell the other thread we're done. */
-  if (cmpxchgi(&handle->pending, 1, 2) != 1)
+  if (cmpxchgi(&handle->u.fd, 1, 0) != 1)
     abort();
 
   return 0;
 }
 
 
 /* Only call this from the event loop thread. */
-static int uv__async_spin(uv_async_t* handle) {
-  int rc;
-
+static void uv__async_spin(uv_async_t* handle) {
   for (;;) {
-    /* rc=0 -- handle is not pending.
-     * rc=1 -- handle is pending, other thread is still working with it.
-     * rc=2 -- handle is pending, other thread is done.
-     */
-    rc = cmpxchgi(&handle->pending, 2, 0);
-
-    if (rc != 1)
-      return rc;
+    /* mark the handle to be closed if not busy */
+    if (cmpxchgi(&handle->u.fd, 0, 2) == 0)
+        break;
```

Comment:
```suggestion
      break;
```
And since there's only one caller of `uv__async_spin()` left you might as well inline it in `uv__async_close()`.